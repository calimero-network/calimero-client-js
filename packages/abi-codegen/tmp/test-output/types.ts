/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */

export interface AbiState {
  counters: Record<string, number>;
  users: UserId32[];
}

export type Action =
  | { kind: "Ping" }
  | { kind: "SetName"; payload: string }
  | { kind: "Update"; payload: UpdatePayload }

export type ConformanceError =
  | { kind: "BadInput" }
  | { kind: "NotFound"; payload: string }

/** Fixed-length bytes (size: 64). Represented as Uint8Array at runtime. */
export type Hash64 = Uint8Array;

export interface Person {
  id: UserId32;
  name: string;
  age: number;
}

export interface Profile {
  bio: string | null;
  avatar: Uint8Array | null;
  nicknames: string[];
}

/** Fixed-length bytes (size: 32). Represented as Uint8Array at runtime. */
export type UserId32 = Uint8Array;

export interface UpdatePayload {
  id: UserId32;
  data: string;
}

export type may_failErrorCode = "BAD_INPUT" | "NOT_FOUND";
export type may_failError = { code: may_failErrorCode } & (
  | { code: "BAD_INPUT" }
  | { code: "NOT_FOUND"; payload: string }
);

export type find_personErrorCode = "BAD_INPUT" | "NOT_FOUND";
export type find_personError = { code: find_personErrorCode } & (
  | { code: "BAD_INPUT" }
  | { code: "NOT_FOUND"; payload: string }
);


export type NamedPayload = string;

export type DataPayload = Uint8Array;

export type PersonUpdatedPayload = string;

export type ActionTakenPayload = string;

export type AbiEvent =
  | { name: "Ping" }
  | { name: "Named"; payload: string }
  | { name: "Data"; payload: Uint8Array }
  | { name: "PersonUpdated"; payload: string }
  | { name: "ActionTaken"; payload: string }
;
