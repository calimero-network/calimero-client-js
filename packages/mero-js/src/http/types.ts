import { ResponseData } from '../types/api-response';

// Web Standards based types
export type FetchLike = (
  input: RequestInfo,
  init?: RequestInit,
) => Promise<Response>;

export interface Transport {
  fetch: FetchLike;
  baseUrl: string;
  defaultHeaders?: Record<string, string>;
  getAuthToken?: () => Promise<string | undefined>;
  onTokenRefresh?: (newToken: string) => Promise<void>;
  timeoutMs?: number;
}

export interface HttpClient {
  get<T>(path: string, init?: RequestInit): Promise<ResponseData<T>>;
  post<T>(
    path: string,
    body?: unknown,
    init?: RequestInit,
  ): Promise<ResponseData<T>>;
  put<T>(
    path: string,
    body?: unknown,
    init?: RequestInit,
  ): Promise<ResponseData<T>>;
  delete<T>(path: string, init?: RequestInit): Promise<ResponseData<T>>;
  patch<T>(
    path: string,
    body?: unknown,
    init?: RequestInit,
  ): Promise<ResponseData<T>>;
  head<T>(path: string, init?: RequestInit): Promise<ResponseData<T>>;
  request<T>(path: string, init?: RequestInit): Promise<ResponseData<T>>;
}

// Legacy compatibility types (for gradual migration)
export interface Header {
  [key: string]: string;
}

export interface ProgressCallback {
  (progress: number): void;
}

export interface HeadResponse {
  headers: Record<string, string>;
  status: number;
}

export interface RequestOptions {
  responseType?: 'arraybuffer' | 'blob' | 'json';
}

// Note: Auth and storage interfaces moved to legacy-types.ts for Axios compatibility
