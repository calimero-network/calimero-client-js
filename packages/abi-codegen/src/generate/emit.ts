/**
 * Helper functions for emitting generated code
 */

/**
 * Generate a file banner for generated files
 */
export function generateFileBanner(): string {
  return '/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */\n';
}

/**
 * Format a TypeScript identifier to be safe for use in code
 * @param name - The name to format
 * @returns A safe TypeScript identifier
 */
export function formatIdentifier(name: string): string {
  // Handle reserved keywords and special characters
  const reservedKeywords = new Set([
    'break',
    'case',
    'catch',
    'class',
    'const',
    'continue',
    'debugger',
    'default',
    'delete',
    'do',
    'else',
    'enum',
    'export',
    'extends',
    'false',
    'finally',
    'for',
    'function',
    'if',
    'import',
    'in',
    'instanceof',
    'let',
    'new',
    'null',
    'return',
    'super',
    'switch',
    'this',
    'throw',
    'true',
    'try',
    'typeof',
    'var',
    'void',
    'while',
    'with',
    'yield',
    'abstract',
    'boolean',
    'byte',
    'char',
    'double',
    'final',
    'float',
    'goto',
    'implements',
    'int',
    'interface',
    'long',
    'native',
    'package',
    'private',
    'protected',
    'public',
    'short',
    'static',
    'synchronized',
    'throws',
    'transient',
    'volatile',
  ]);

  if (reservedKeywords.has(name)) {
    return `${name}_`;
  }

  // Replace invalid characters with underscores
  return name.replace(/[^a-zA-Z0-9_$]/g, '_');
}

/**
 * Convert a string to PascalCase for type names
 * @param str - The string to convert
 * @returns PascalCase string
 */
export function toPascalCase(str: string): string {
  return str
    .split(/[-_\s]+/)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join('');
}

/**
 * Convert a string to camelCase for method names
 * @param str - The string to convert
 * @returns camelCase string
 */
export function toCamelCase(str: string): string {
  const pascal = toPascalCase(str);
  return pascal.charAt(0).toLowerCase() + pascal.slice(1);
}

/**
 * Derive a client class name from a file path
 * @param p - The file path to derive the name from
 * @returns The derived client class name
 */
export function deriveClientNameFromPath(p: string): string {
  const base = (p.split(/[\\/]/).pop() || "").replace(/\.[^.]+$/, ""); // strip extension
  const parts = base.split(/[^A-Za-z0-9]+/).filter(Boolean);
  const word = (s: string) =>
    s.length <= 2 ? s.toUpperCase() : (s[0].toUpperCase() + s.slice(1).toLowerCase());
  const pascal = parts.map(word).join("");
  return pascal + "Client";
}
