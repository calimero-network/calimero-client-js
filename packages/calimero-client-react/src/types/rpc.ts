import { ContextId } from './context';

export type RpcRequestId = string | number;

export interface RpcClient {
  execute<Args, Out>(
    params: RpcQueryParams<Args>,
    config?: RequestConfig,
  ): Promise<RpcResult<RpcQueryResponse<Out>>>;
}

interface Headers {
  [key: string]: string;
}

export interface RequestConfig {
  timeout?: number;
  headers?: Headers;
}

export type RpcResult<Result> =
  | {
      result: Result;
      error?: null;
    }
  | {
      result?: null;
      error: RpcError;
    };

export interface RpcErrorInfo {
  name: string;
  cause: {
    name: string;
    info?: {
      message: string;
    };
  };
}

export interface RpcError {
  id: RpcRequestId;
  jsonrpc: string;
  code?: number;
  error: RpcErrorInfo;
  headers?: {
    [key: string]: string;
  };
}

export interface RpcQueryParams<Args> {
  contextId: ContextId;
  method: string;
  argsJson: Args;
  executorPublicKey: string;
}

export interface RpcQueryResponse<Output> {
  output?: Output;
}
