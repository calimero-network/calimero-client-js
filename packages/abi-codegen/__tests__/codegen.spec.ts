import { describe, it, expect, beforeAll } from 'vitest';
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import { loadAbiManifestFromFile } from '../src/parse.js';
import { generateTypes } from '../src/generate/types.js';
import { generateClient } from '../src/generate/client.js';
import { deriveClientNameFromPath } from '../src/generate/emit.js';

describe('Codegen', () => {
  const conformanceAbiPath = path.join(
    __dirname,
    '../__fixtures__/abi_conformance.json',
  );
  let manifest: any;

  beforeAll(() => {
    manifest = loadAbiManifestFromFile(conformanceAbiPath);
  });

  describe('types.ts generation', () => {
    it('should generate types.ts with correct structure', () => {
      const typesContent = generateTypes(manifest);

      // Snapshot the generated types
      expect(typesContent).toMatchSnapshot();

      // Assert key patterns
      expect(typesContent).toContain('export interface Person {');
      expect(typesContent).toContain('export type ActionPayload =');
      expect(typesContent).toContain("| { name: 'SetName'; payload: string }");
      expect(typesContent).toContain(
        "| { name: 'Update'; payload: UpdatePayload }",
      );
      expect(typesContent).toContain('export type UserId32 = string;');
      expect(typesContent).toContain(
        '/** Fixed-length bytes (size: 32). Represented as string at runtime. */',
      );
      expect(typesContent).toContain(
        'export type may_failErrorCode = "BAD_INPUT" | "NOT_FOUND";',
      );
      expect(typesContent).toContain('export type AbiEvent =');
      expect(typesContent).toContain(
        '| { name: "PersonUpdated"; payload: string }',
      );
      // Unit events should not have payload property
      expect(typesContent).toContain('| { name: "Ping" }');
      expect(typesContent).not.toContain('| { name: "Ping"; payload:');
    });

    it('should handle nullable fields correctly', () => {
      const typesContent = generateTypes(manifest);

      expect(typesContent).toContain('bio: string | null;');
      expect(typesContent).toContain('bio: string | null;');
    });

    it('should handle maps correctly', () => {
      const typesContent = generateTypes(manifest);

      expect(typesContent).toContain('counters: Record<string, number>;');
    });

    it('should handle lists correctly', () => {
      const typesContent = generateTypes(manifest);

      // The conformance ABI has list_u32 method that returns number[]
      // But we need to check the client generation for this
      const clientContent = generateClient(manifest);
      expect(clientContent).toContain('number[]');
    });

    it('should handle unit returns correctly', () => {
      const clientContent = generateClient(manifest);

      // The 'act' method returns number
      expect(clientContent).toContain(
        'async act(params: { a: Types.ActionPayload }): Promise<number> {',
      );
    });

    it('should handle unit events correctly', () => {
      const typesContent = generateTypes(manifest);

      // Unit events should not have payload property in the union
      expect(typesContent).toContain('| { name: "Ping" }');
      expect(typesContent).not.toContain('| { name: "Ping"; payload:');

      // Unit events should not generate payload type aliases
      expect(typesContent).not.toContain('export type PingPayload =');
    });

    it('should include fixed bytes JSDoc', () => {
      const typesContent = generateTypes(manifest);

      // Fixed bytes should have JSDoc with size information
      expect(typesContent).toContain(
        '/** Fixed-length bytes (size: 32). Represented as string at runtime. */',
      );
      expect(typesContent).toContain(
        '/** Fixed-length bytes (size: 64). Represented as string at runtime. */',
      );
    });

    it('should handle variant inline struct payloads correctly', () => {
      const typesContent = generateTypes(manifest);

      // The Action variant "Update" has a reference to UpdatePayload
      expect(typesContent).toContain(
        "| { name: 'Update'; payload: UpdatePayload }",
      );
    });

    it('should handle variant $ref payloads correctly', () => {
      const typesContent = generateTypes(manifest);

      // The Action variant "SetName" has a string payload
      expect(typesContent).toContain("| { name: 'SetName'; payload: string }");
      // The Action variant "Ping" has no payload
      expect(typesContent).toContain("| { name: 'Ping' }");
    });
  });

  describe('client.ts generation', () => {
    it('should generate client.ts with correct structure', () => {
      const clientContent = generateClient(manifest, 'TestClient');

      // Snapshot the generated client
      expect(clientContent).toMatchSnapshot();

      // Assert key patterns
      expect(clientContent).toContain('export class TestClient {');
      expect(clientContent).toContain('import {');
      expect(clientContent).toContain('  CalimeroApp,');
      expect(clientContent).toContain('  Context,');
      // ExecutionResponse is no longer imported
      expect(clientContent).toContain(
        "} from '@calimero-network/calimero-client';",
      );
      expect(clientContent).toContain(
        'constructor(app: CalimeroApp, context: Context) {',
      );
      expect(clientContent).toContain(
        'async optU32(params: { x: number | null }): Promise<number | null> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'opt_u32', params);",
      );
      expect(clientContent).toContain(
        'async makePerson(params: { p: Types.Person }): Promise<Types.Person> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'make_person', convertedParams);",
      );
      expect(clientContent).toContain(
        '@throws {Types.may_failError} May throw the following errors:',
      );
      expect(clientContent).toContain('- BAD_INPUT');
      expect(clientContent).toContain('- NOT_FOUND: string');
    });

    it('should handle zero-parameter methods correctly', () => {
      // Note: The test ABI doesn't have any zero-parameter methods
      // This test would verify that zero-parameter methods have no arguments and pass empty object
      // For now, we'll skip this test since our test ABI has no such methods
      expect(true).toBe(true);
    });

    it('should handle single-parameter methods correctly', () => {
      const clientContent = generateClient(manifest);

      // Check that single-parameter methods use named params object
      expect(clientContent).toContain(
        'async roundtripId(params: { x: string }): Promise<string> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'roundtrip_id', convertedParams);",
      );
      expect(clientContent).toContain(
        'async optU32(params: { x: number | null }): Promise<number | null> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'opt_u32', params);",
      );
    });

    it('should handle multi-parameter methods correctly', () => {
      const clientContent = generateClient(manifest);

      // Check that multi-parameter methods use named params object with all fields
      expect(clientContent).toContain(
        'async makePerson(params: { p: Types.Person }): Promise<Types.Person> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'make_person', convertedParams);",
      );
    });

    it('should preserve ABI method names verbatim in execute calls', () => {
      const clientContent = generateClient(manifest);

      // Check that ABI method names are used exactly as they appear in the ABI
      expect(clientContent).toContain("'opt_u32'");
      expect(clientContent).toContain("'list_u32'");
      expect(clientContent).toContain("'map_u32'");
      expect(clientContent).toContain("'make_person'");
      expect(clientContent).toContain("'profile_roundtrip'");
      expect(clientContent).toContain("'roundtrip_id'");
      expect(clientContent).toContain("'roundtrip_hash'");
      expect(clientContent).toContain("'may_fail'");
      expect(clientContent).toContain("'find_person'");
    });

    it('should handle methods with errors correctly', () => {
      const clientContent = generateClient(manifest);

      expect(clientContent).toContain(
        'async mayFail(params: { flag: boolean }): Promise<number> {',
      );
      expect(clientContent).toContain('- NOT_FOUND: string');
    });

    it('should handle methods with nullable returns', () => {
      const clientContent = generateClient(manifest);

      expect(clientContent).toContain(
        'async findPerson(params: { name: string }): Promise<Types.Person> {',
      );
    });

    it('should preserve parameter order in method calls', () => {
      const clientContent = generateClient(manifest);

      // The makePerson method has a single parameter of type Person
      expect(clientContent).toContain(
        'async makePerson(params: { p: Types.Person }): Promise<Types.Person> {',
      );
      expect(clientContent).toContain(
        "const response = await this.app.execute(this.context, 'make_person', convertedParams);",
      );
    });

    it('should include barrel export from types', () => {
      const clientContent = generateClient(manifest);

      // Should re-export all types from types.ts
      expect(clientContent).toContain('export * from "./types.js";');
    });

    it('should include generated banner', () => {
      const clientContent = generateClient(manifest);
      const typesContent = generateTypes(manifest);

      // Both files should have the generated banner
      expect(clientContent).toContain(
        '/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */',
      );
      expect(typesContent).toContain(
        '/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */',
      );
    });
  });

  describe('name derivation', () => {
    it('should derive client name from wasm filename correctly', () => {
      expect(deriveClientNameFromPath('kv_store.wasm')).toBe('KVStoreClient');
      expect(deriveClientNameFromPath('plantr.wasm')).toBe('PlantrClient');
      expect(deriveClientNameFromPath('abi-conformance.wasm')).toBe(
        'AbiConformanceClient',
      );
    });

    it('should derive client name from json filename correctly', () => {
      expect(deriveClientNameFromPath('abi-conformance.json')).toBe(
        'AbiConformanceClient',
      );
      expect(deriveClientNameFromPath('kv_store.json')).toBe('KVStoreClient');
    });

    it('should handle paths with directories', () => {
      expect(deriveClientNameFromPath('/tmp/kv_store.wasm')).toBe(
        'KVStoreClient',
      );
      expect(
        deriveClientNameFromPath(
          'apps/kv_store/target/wasm32-unknown-unknown/debug/kv_store.wasm',
        ),
      ).toBe('KVStoreClient');
    });

    it('should handle edge cases', () => {
      expect(deriveClientNameFromPath('')).toBe('Client');
      expect(deriveClientNameFromPath('a')).toBe('AClient');
      expect(deriveClientNameFromPath('ab')).toBe('ABClient');
      expect(deriveClientNameFromPath('abc')).toBe('AbcClient');
    });

    it('should generate client with derived name', () => {
      const clientContent = generateClient(manifest, 'KVStoreClient');

      expect(clientContent).toContain('export class KVStoreClient {');
      expect(clientContent).toContain(
        'constructor(app: CalimeroApp, context: Context) {',
      );
      expect(clientContent).toContain('import {');
      expect(clientContent).toContain('  CalimeroApp,');
      expect(clientContent).toContain('  Context,');
      expect(clientContent).toContain(
        "} from '@calimero-network/calimero-client';",
      );
    });
  });

  describe('manifest immutability', () => {
    it('should not mutate the manifest during generation', () => {
      // Test that the manifest is deeply frozen
      expect(() => {
        // This should throw if the manifest is properly frozen
        (manifest as any).methods.push({ name: 'test', params: [] });
      }).toThrow();

      // Test a nested variant payload node
      expect(() => {
        // This should throw if the manifest is properly frozen
        (manifest.types.Action.variants[1] as any).payload = { kind: 'string' };
      }).toThrow();
    });
  });

  describe('compile test', () => {
    it('should generate code that compiles under tsc --strict', () => {
      const typesContent = generateTypes(manifest);
      const clientContent = generateClient(manifest);

      // Create a temporary directory for the test
      const tmpDir = path.join(__dirname, '../tmp/gen');
      if (!fs.existsSync(tmpDir)) {
        fs.mkdirSync(tmpDir, { recursive: true });
      }

      // Write the generated files
      const typesPath = path.join(tmpDir, 'types.ts');
      const clientPath = path.join(tmpDir, 'client.ts');
      fs.writeFileSync(typesPath, typesContent);
      fs.writeFileSync(clientPath, clientContent);

      // Update client.ts to import from the same directory
      const updatedClientContent = clientContent.replace(
        'export * from "./types.js";',
        'export * from "./types";',
      );
      fs.writeFileSync(clientPath, updatedClientContent);

      // Also need to include the types directly in client.ts for the compile test
      const clientWithTypes = updatedClientContent.replace(
        'export * from "./types";',
        '// Include types directly for compile test\n' + typesContent,
      );

      // Remove the calimero-client import and add mock types for compile test
      const clientWithMockedImport = clientWithTypes.replace(
        `import {
  CalimeroApp,
  Context,
} from '@calimero-network/calimero-client';`,
        `// Mock types for compile test
type CalimeroApp = any;
type Context = any;`,
      );

      // Add a test stub to verify the new parameter structure compiles
      const testStub = `
// Test stub to verify parameter structure
const app = { execute: async (_c: any, _m: string, _p?: Record<string, unknown>) => ({ success: true, result: null }) } as any;
const ctx = {} as any;
const client = new Client(app, ctx);
await client.roundtripId({ x: "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef" });
await client.makePerson({ p: { id: "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef" as any, name: "test", age: 25 } });
await client.act({ a: Action.Ping() });
`;
      const clientWithTestStub = clientWithMockedImport + testStub;
      fs.writeFileSync(clientPath, clientWithTestStub);

      // Create a minimal tsconfig.json
      const tsconfigPath = path.join(tmpDir, 'tsconfig.json');
      const tsconfig = {
        compilerOptions: {
          target: 'ES2020',
          module: 'ESNext',
          moduleResolution: 'node',
          strict: true,
          noEmit: true,
          esModuleInterop: true,
          skipLibCheck: true,
          forceConsistentCasingInFileNames: true,
        },
        include: ['*.ts'],
      };
      fs.writeFileSync(tsconfigPath, JSON.stringify(tsconfig, null, 2));

      // Run tsc --strict --noEmit using the local TypeScript installation
      try {
        execSync('npx tsc --strict --noEmit', {
          cwd: tmpDir,
          stdio: 'pipe',
          encoding: 'utf-8',
        });
      } catch (error: any) {
        // If compilation fails, show the error
        console.error('TypeScript compilation failed:');
        console.error(error.stdout || error.stderr);
        throw error;
      }
    });
  });
});
