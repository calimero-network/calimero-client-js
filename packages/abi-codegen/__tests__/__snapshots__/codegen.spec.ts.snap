// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Codegen > client.ts generation > should generate client.ts with correct structure 1`] = `
"/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */

import {
  CalimeroApp,
  Context,
} from '@calimero-network/calimero-client';

import * as Types from "./types.js";

export * from "./types.js";

export class TestClient {
  private app: CalimeroApp;
  private context: Context;

  constructor(app: CalimeroApp, context: Context) {
    this.app = app;
    this.context = context;
  }

  /**
   * Utility function to convert hex string to Uint8Array
   */
  private hexToBytes(hex: string): Uint8Array {
    return new Uint8Array(hex.match(/.{1,2}/g)?.map(byte => parseInt(byte, 16)) || []);
  }

  /**
   * Utility function to convert Uint8Array to hex string
   */
  private bytesToHex(bytes: Uint8Array): string {
    return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
  }

  /**
   * Utility function to deserialize bytes from contract responses
   */
  private deserializeBytes(result: any): Uint8Array {
    if (Array.isArray(result)) {
      return new Uint8Array(result);
    }
    // Handle case where result might be a hex string
    if (typeof result === 'string') {
      return this.hexToBytes(result);
    }
    throw new Error(\`Unexpected bytes result format: \${typeof result}\`);
  }

  /**
   * init
   */
  public async init(): Promise<Types.AbiState> {
    const response = await this.app.execute(this.context, 'init', {});
    if (response.success) {
      return response.result as Types.AbiState;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * noop
   */
  public async noop(): Promise<void> {
    const response = await this.app.execute(this.context, 'noop', {});
    if (response.success) {
      return response.result as void;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_bool
   */
  public async echoBool(params: { b: boolean }): Promise<boolean> {
    const response = await this.app.execute(this.context, 'echo_bool', params);
    if (response.success) {
      return response.result as boolean;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_i32
   */
  public async echoI32(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_i32', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_i64
   */
  public async echoI64(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_i64', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_u32
   */
  public async echoU32(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_u32', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_u64
   */
  public async echoU64(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_u64', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_f32
   */
  public async echoF32(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_f32', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_f64
   */
  public async echoF64(params: { v: number }): Promise<number> {
    const response = await this.app.execute(this.context, 'echo_f64', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_string
   */
  public async echoString(params: { s: string }): Promise<string> {
    const response = await this.app.execute(this.context, 'echo_string', params);
    if (response.success) {
      return response.result as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * echo_bytes
   */
  public async echoBytes(params: { b: string }): Promise<string> {
    const convertedParams = {
      b: Array.from(this.hexToBytes(params.b)),
    };
    const response = await this.app.execute(this.context, 'echo_bytes', convertedParams);
    if (response.success) {
      const resultBytes = this.deserializeBytes(response.result);
      return this.bytesToHex(resultBytes) as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * opt_u32
   */
  public async optU32(params: { x: number | null }): Promise<number | null> {
    const response = await this.app.execute(this.context, 'opt_u32', params);
    if (response.success) {
      return response.result as number | null;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * opt_string
   */
  public async optString(params: { x: string | null }): Promise<string | null> {
    const response = await this.app.execute(this.context, 'opt_string', params);
    if (response.success) {
      return response.result as string | null;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * opt_record
   */
  public async optRecord(params: { p: Types.Person | null }): Promise<Types.Person | null> {
    const response = await this.app.execute(this.context, 'opt_record', params);
    if (response.success) {
      return response.result as Types.Person | null;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * opt_id
   */
  public async optId(params: { x: string | null }): Promise<string | null> {
    const convertedParams = {
      x: params.x ? Array.from(this.hexToBytes(params.x)) : null,
    };
    const response = await this.app.execute(this.context, 'opt_id', convertedParams);
    if (response.success) {
      if (response.result === null) {
        return null as string | null;
      }
      const resultBytes = this.deserializeBytes(response.result);
      return this.bytesToHex(resultBytes) as string | null;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * list_u32
   */
  public async listU32(params: { xs: number[] }): Promise<number[]> {
    const response = await this.app.execute(this.context, 'list_u32', params);
    if (response.success) {
      return response.result as number[];
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * list_strings
   */
  public async listStrings(params: { xs: string[] }): Promise<string[]> {
    const response = await this.app.execute(this.context, 'list_strings', params);
    if (response.success) {
      return response.result as string[];
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * list_records
   */
  public async listRecords(params: { ps: string }): Promise<string> {
    const response = await this.app.execute(this.context, 'list_records', params);
    if (response.success) {
      return response.result as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * list_ids
   */
  public async listIds(params: { xs: string }): Promise<string> {
    const response = await this.app.execute(this.context, 'list_ids', params);
    if (response.success) {
      return response.result as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * map_u32
   */
  public async mapU32(params: { m: Record<string, number> }): Promise<Record<string, number>> {
    const response = await this.app.execute(this.context, 'map_u32', params);
    if (response.success) {
      return response.result as Record<string, number>;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * map_list_u32
   */
  public async mapListU32(params: { m: Record<string, number[]> }): Promise<Record<string, number[]>> {
    const response = await this.app.execute(this.context, 'map_list_u32', params);
    if (response.success) {
      return response.result as Record<string, number[]>;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * map_record
   */
  public async mapRecord(params: { m: string }): Promise<string> {
    const response = await this.app.execute(this.context, 'map_record', params);
    if (response.success) {
      return response.result as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * make_person
   */
  public async makePerson(params: { p: Types.Person }): Promise<Types.Person> {
    const response = await this.app.execute(this.context, 'make_person', params);
    if (response.success) {
      return response.result as Types.Person;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * profile_roundtrip
   */
  public async profileRoundtrip(params: { p: Types.Profile }): Promise<Types.Profile> {
    const response = await this.app.execute(this.context, 'profile_roundtrip', params);
    if (response.success) {
      return response.result as Types.Profile;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * act
   */
  public async act(params: { a: Types.Action }): Promise<number> {
    const response = await this.app.execute(this.context, 'act', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * roundtrip_id
   */
  public async roundtripId(params: { x: string }): Promise<string> {
    const convertedParams = {
      x: Array.from(this.hexToBytes(params.x)),
    };
    const response = await this.app.execute(this.context, 'roundtrip_id', convertedParams);
    if (response.success) {
      const resultBytes = this.deserializeBytes(response.result);
      return this.bytesToHex(resultBytes) as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * roundtrip_hash
   */
  public async roundtripHash(params: { h: string }): Promise<string> {
    const convertedParams = {
      h: Array.from(this.hexToBytes(params.h)),
    };
    const response = await this.app.execute(this.context, 'roundtrip_hash', convertedParams);
    if (response.success) {
      const resultBytes = this.deserializeBytes(response.result);
      return this.bytesToHex(resultBytes) as string;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * may_fail
   *
   * @throws {Error} May throw the following errors:
   * - BAD_INPUT
   * - NOT_FOUND: string
   */
  public async mayFail(params: { flag: boolean }): Promise<number> {
    const response = await this.app.execute(this.context, 'may_fail', params);
    if (response.success) {
      return response.result as number;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

  /**
   * find_person
   *
   * @throws {Error} May throw the following errors:
   * - BAD_INPUT
   * - NOT_FOUND: string
   */
  public async findPerson(params: { name: string }): Promise<Types.Person> {
    const response = await this.app.execute(this.context, 'find_person', params);
    if (response.success) {
      return response.result as Types.Person;
    } else {
      throw new Error(response.error || 'Execution failed');
    }
  }

}"
`;

exports[`Codegen > types.ts generation > should generate types.ts with correct structure 1`] = `
"/** @generated by @calimero/abi-codegen â€” DO NOT EDIT. */

export interface AbiState {
  counters: Record<string, number>;
  users: UserId32[];
}

export type Action =
  | { kind: "Ping" }
  | { kind: "SetName"; payload: string }
  | { kind: "Update"; payload: UpdatePayload }

export type ConformanceError =
  | { kind: "BadInput" }
  | { kind: "NotFound"; payload: string }

/** Fixed-length bytes (size: 64). Represented as Uint8Array at runtime. */
export type Hash64 = Uint8Array;

export interface Person {
  id: UserId32;
  name: string;
  age: number;
}

export interface Profile {
  bio: string | null;
  avatar: Uint8Array | null;
  nicknames: string[];
}

/** Fixed-length bytes (size: 32). Represented as Uint8Array at runtime. */
export type UserId32 = Uint8Array;

export interface UpdatePayload {
  id: UserId32;
  data: string;
}

export type may_failErrorCode = "BAD_INPUT" | "NOT_FOUND";
export type may_failError = { code: may_failErrorCode } & (
  | { code: "BAD_INPUT" }
  | { code: "NOT_FOUND"; payload: string }
);

export type find_personErrorCode = "BAD_INPUT" | "NOT_FOUND";
export type find_personError = { code: find_personErrorCode } & (
  | { code: "BAD_INPUT" }
  | { code: "NOT_FOUND"; payload: string }
);


export type NamedPayload = string;

export type DataPayload = Uint8Array;

export type PersonUpdatedPayload = string;

export type ActionTakenPayload = string;

export type AbiEvent =
  | { name: "Ping" }
  | { name: "Named"; payload: string }
  | { name: "Data"; payload: Uint8Array }
  | { name: "PersonUpdated"; payload: string }
  | { name: "ActionTaken"; payload: string }
;
"
`;
